### PR TITLE
Parameterless struct constructors: Remaining work items

### PR DESCRIPTION
**Pull request by jcouv**
This is addressing remaining work items in test plan: https://github.com/dotnet/roslyn/issues/51698

See spec for expected execution of field initializers and `: this()` representing a default value type constructor:
https://github.com/dotnet/csharplang/blob/main/proposals/parameterless-struct-constructors.md#executing-field-initializers

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -3388,39 +3388,63 @@ internal virtual BoundExpressionStatement BindConstructorInitializer(PrimaryCons
...
                 if (!constructorSymbol.IsStatic &&
                     !SynthesizedRecordCopyCtor.IsCopyConstructor(constructorSymbol))
                 {
                     // Note: we check the constructor initializer of copy constructors elsewhere
-                    Error(diagnostics, ErrorCode.ERR_UnexpectedOrMissingConstructorInitializerInRecord, constructor.Initializer?.ThisOrBaseKeyword ?? constructor.Identifier);
+                    Error(diagnostics, ErrorCode.ERR_UnexpectedOrMissingConstructorInitializerInRecord, initializer?.ThisOrBaseKeyword ?? constructor.Identifier);
                 }
             }
 
+            bool skipInitializer = false;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:3415]: >false [](http://example.com/codeflow?start=35&length=5)

Consider inlining the `if` condition.

Also

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 2]
```diff
@@ -3388,39 +3388,63 @@ internal virtual BoundExpressionStatement BindConstructorInitializer(PrimaryCons
...
+                skipInitializer = true;
+            }
+
             // Using BindStatement to bind block to make sure we are reusing results of partial binding in SemanticModel
             return new BoundConstructorMethodBody(constructor,
                                                   bodyBinder.GetDeclaredLocalsForScope(constructor),
-                                                  constructor.Initializer == null ? null : bodyBinder.BindConstructorInitializer(constructor.Initializer, diagnostics),
+                                                  skipInitializer ? new BoundNoOpStatement(constructor, NoOpStatementFlavor.Default)
+                                                      : initializer == null ? null
+                                                      : bodyBinder.BindConstructorInitializer(initializer, diagnostics),
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:3430]: >: [](http://example.com/codeflow?start=54&length=1)

Should this be indented further, since it's associated with the nested conditional expression?

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 3]
```diff
@@ -3388,39 +3388,63 @@ internal virtual BoundExpressionStatement BindConstructorInitializer(PrimaryCons
...
                                                   constructor.Body == null ? null : (BoundBlock)bodyBinder.BindStatement(constructor.Body, diagnostics),
                                                   constructor.ExpressionBody == null ?
                                                       null :
                                                       bodyBinder.BindExpressionBodyAsBlock(constructor.ExpressionBody,
                                                                                            constructor.Body == null ? diagnostics : BindingDiagnosticBag.Discarded));
         }
 
+        internal static bool IsDefaultValueTypeConstructor(NamedTypeSymbol type, ConstructorInitializerSyntax initializerSyntax)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:3438]: >IsDefaultValueTypeConstructor [](http://example.com/codeflow?start=29&length=29)

Consider moving back to `MemberSymbolExtensions`, perhaps as an extension method on `NamedTypeSymbol type`, since this is not related to the `Binder`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 4]
```diff
@@ -1048,6 +1048,77 @@ public void TypeDeclaration_ParameterlessConstructor_02()
...
+}
+record struct S3()
+{
+    public string field = ""hello"";
+}
+";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "(hello, hello, True, hello)");
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:1120]: >} [](http://example.com/codeflow?start=8&length=1)

Are we testing calls to the following constructors?
```csharp
record struct S4
{
    public string field = "hello";
    public S4(object o) : this() { }
}
record struct S5()
{
    public string field = "hello";
    public S5(object o) : this() { }
}
record struct S6(string other)
{
    public string field = "hello";
    public S6() : this(null) { }
}
```

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:3430]: I think that's how we've indented such nested conditionals before.

**Comment by jcouv** [comment]
@333fred @dotnet/roslyn-compiler for second review. Thanks

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:3426]: If it is, then when I next stumble onto those cases I'll reformat for readability. The way this is formatted is very confusing.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 5]
```diff
@@ -1048,6 +1048,96 @@ public void TypeDeclaration_ParameterlessConstructor_02()
...
+record struct S6(object o)
+{
+    public S6() : this(null) { }
+}
+";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (13,5): error CS8862: A constructor declared in a record with parameter list must have 'this' constructor initializer.
+                //     S3(object o) { } // 1
+                Diagnostic(ErrorCode.ERR_UnexpectedOrMissingConstructorInitializerInRecord, "S3").WithLocation(13, 5),
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:1085]: Can we improve this error message? I don't understand what it's telling me.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:1085]: I suppose this is probably out of scope for this PR, but I really do find this confusing.

**Review by 333fred — APPROVED** [review]
LGTM (commit 5) with ternary formatting fix.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs (hunk 5 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:1085]: As I mentioned offline, this error message correctly describes the problem. But we can certainly improve the wording separately from the parameterless struct ctor feature work if you have a suggestion.
